File: Aeds/AedsII2/Trabalhos/Tp2/Q03.java
import java.util.*;
import java.io.FileReader;
import java.io.BufferedReader;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

/**
 * Questao 3 - Trabalho Pratico 2 - Aeds II - CC_PUCMINAS
 * @author Tulio Gomes Braga
 * @version 2 04/2025
 * 
 * 
 * Explicações dos Métodos:
 * 
 * Personagem: Responsavel pela cricao dos atributos e operacoes especificas de personagem.
 *  Personagem(): Metodo construtor sem parametro.
 *  Personegem(parametros): Metodo construtor com parametro.
 *  Metodos set: Responsaveis por receber o parmetro e atribuir o mesmo ao atribuito.
 *  Metodos get: Responsaveis por retornar o atributo. 
 *  Clone: Copia os dados de outro personagem para um novo.
 *  mostrarPersonagem: Apresenta na saida padrao os atributos formatados.
 * 
 * CtrlPersonagem: Responsavel pelas operacoes envovendo personagens.
 *  setaBase: Abre o arquivo uma vez e passa linha por linha para setar todos os personagens.
 *  setaPersonagem: Particiona a linha e chama os metodos sets correspondentes aos atributos.
 *  mostrarBase: Percorre toda a base fazendo objeto.mostrarPersonagem().
 *  setaLista: Equanto a entrada for diferente de "FIM", chama-se o medoto setaInput(passando a entrada lida).
 *  mostraLista: Percorre toda a lista fazendo objeto.mostrarPersonagem().
 * 
 *  pesquisaSequencial: Percorre toda lista procurando a chave de pesquisa, apos isso apresenta o resultado da busca.
 */

class Personagem {

    private String id, name, house, ancestry, species, patronus, hogwartsStaff, hogwartsStudent, actorName, eyeColour, gender, hairColour;
    private String[] alternate_names;
    private String[] alternate_actors;
    private LocalDate dateOfBirth;
    private int yearOfBirth;
    private boolean wizard, alive;


    public Personagem() {
        this.id = null;
        this.name = null;
        this.house = null;
        this.ancestry = null;
        this.species = null;
        this.patronus = null;
        this.hogwartsStaff = null;
        this.hogwartsStudent = null;
        this.actorName = null;
        this.eyeColour = null;
        this.gender = null;
        this.hairColour = null;
        this.alternate_names = new String[10];
        this.alternate_actors = new String[10];
        this.dateOfBirth = null;
        this.yearOfBirth = 0;
        this.wizard = false;
        this.alive = false;
    }

    public Personagem(String id, String name, String house, String ancestry, String species, String patronus, String hogwartsStaff, String hogwartsStudent, String actorName, String eyeColour, String gender, String hairColour, String[] alternate_names, String[] alternate_actors, LocalDate dateOfBirth, int yearOfBirth, boolean alive, boolean wizard) {
        this.id = id;
        this.name = name;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.alternate_names = alternate_names;
        this.alternate_actors = alternate_actors;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.alive = alive;
        this.wizard = wizard;
    }

    public void setId(String id) {this.id = id;}
    public String getId() {return id;}

    public void setName(String name) {this.name = name;}
    public String getName() {return name;}

    public void setHouse(String house) {this.house = house;}
    public String getHouse() {return house;}

    public void setAncestry(String ancestry) {this.ancestry = ancestry;}
    public String getAncestry() {return ancestry;}

    public void setSpecies(String species) {this.species = species;}
    public String getSpecies() {return species;}

    public void setPatronus(String patronus) {this.patronus = patronus;}
    public String getPatronus() {return patronus;}

    public void setHogwartsStaff(String hogwartsStaff) {this.hogwartsStaff = hogwartsStaff;}
    public String getHogwartsStaff() {return hogwartsStaff.equals("VERDADEIRO") ? "true" : "false";}

    public void setHogwartsStudent(String hogwartsStudent) {this.hogwartsStudent = hogwartsStudent;}
    public String getHogwartsStudent() {return hogwartsStudent.equals("VERDADEIRO") ? "true" : "false";}

    public void setActorName(String actorName) {this.actorName = actorName;}
    public String getActorName() {return actorName;}

    public void setEyeColour(String eyeColour) {this.eyeColour = eyeColour;}
    public String getEyeColour() {return eyeColour;}

    public void setGender(String gender) {this.gender = gender;}
    public String getGender() {return gender;}

    public void setHairColour(String hairColour) {this.hairColour = hairColour;}
    public String getHairColour() {return hairColour;}

    public void setYearOfBirth(String yearOfBirth) { if(!yearOfBirth.equals("") && !yearOfBirth.equals("yearOfBirth")) this.yearOfBirth = Integer.parseInt(yearOfBirth);}
    public int getYearOfBirth() {return yearOfBirth;}

    public void setAlive(String alive) {boolean x = false;  if(alive.equals("VERDADEIRO")) x=true;    this.alive = x;}
    public boolean getAlive() {return alive;}

    public void setWizard(String wizard) {boolean x = false; if(wizard.equals("VERDADEIRO")) x=true;  this.wizard = x;}
    public boolean getWizard() {return wizard;}

    public void setDateOfBirth(String dateOfBirth) { 
        if(!dateOfBirth.equals("") && !dateOfBirth.equals("dateOfBirth") && !dateOfBirth.equals("[]")){
            String[] parte = dateOfBirth.split("-");
            this.dateOfBirth = LocalDate.of(Integer.parseInt(parte[2]), Integer.parseInt(parte[1]), Integer.parseInt(parte[0])); 
        }
    }
    public String getDateOfBirth() { DateTimeFormatter f = DateTimeFormatter.ofPattern("dd-MM-yyyy"); String dt = this.dateOfBirth.format(f); return dt;}

    public void setAlternateNames(String alternate_names) {
        if(!alternate_names.equals("[]")){
            alternate_names = alternate_names.replaceAll("[\\'\\[\\]]", ""); //replaceAll("[\\s'\\[\\]]", "");
            String[] parte = alternate_names.split(",");

            for(int i=0; i<parte.length; i++){
                this.alternate_names[i] = parte[i]; System.out.print("");
            }
        }
    }
    public String[] getAlternateNames() {return alternate_names;}

    public void setAlternateActors(String alternate_actors) {
        if(!alternate_actors.equals("[]")){
            alternate_actors = alternate_actors.replaceAll("[\\s'\\[\\]]", "");
            String[] parte = alternate_actors.split(",");

            for(int i=0; i<parte.length; i++){
                this.alternate_actors[i] = parte[i];
            }
        }
    }
    public String[] getAlternateActors() { return alternate_actors;}


    public void clone(Personagem p) {
        this.id = p.id;
        this.name = p.name;
        this.house = p.house;
        this.ancestry = p.ancestry;
        this.species = p.species;
        this.patronus = p.patronus;
        this.hogwartsStaff = p.hogwartsStaff;
        this.hogwartsStudent = p.hogwartsStudent;
        this.actorName = p.actorName;
        this.eyeColour = p.eyeColour;
        this.gender = p.gender;
        this.hairColour = p.hairColour;
        this.alternate_names = p.alternate_names;
        this.alternate_actors = p.alternate_actors;
        this.dateOfBirth = p.dateOfBirth;
        this.yearOfBirth = p.yearOfBirth;
        this.alive = p.alive;
        this.wizard = p.wizard;
    }

    public void mostrarPersonagem(){
        System.out.print("["+ this.id +" ## "+ this.name +" ## {");

        for(int i=0; i<this.alternate_names.length; i++){
            if(this.alternate_names[i]!=null) System.out.print(this.alternate_names[i]);
            else i=this.alternate_names.length;
            if(i<10 && this.alternate_names[i+1]!=null) System.out.print(","); 
        }  

        System.out.print("} ## " + this.house +" ## "+ this.ancestry +" ## "+ this.species +" ## "+ this.patronus +" ## "+ this.getHogwartsStaff() +" ## "+ this.getHogwartsStudent() +" ## "+ this.actorName +" ## "+ this.alive +" ## ");
        /*
        for(int i=0; i<this.alternate_actors.length; i++){
            if(this.alternate_actors[i]!=null) System.out.print(this.alternate_actors[i] +" ## ");
            else i=this.alternate_names.length;
        }    
        */ 
        System.out.println(this.getDateOfBirth() +" ## "+ this.yearOfBirth +" ## "+ this.eyeColour +" ## "+ this.gender +" ## "+ this.hairColour +" ## "+ this.wizard +"]");
    }
}

public class CtrlPersonagem{

    static int tamBase=0, entradas=0, maxTam=405;;
    static Personagem[] Base =  new Personagem[maxTam];
    static Personagem[] Lista = new Personagem[maxTam];


    static void setaBase() throws Exception{
        String linha = "";
        FileReader file = new FileReader("/tmp/characters.csv");
        BufferedReader bf = new BufferedReader(file);

        while((linha=bf.readLine()) != null){
            setaPersonagem(linha);
        }

        bf.close();
        file.close();
    }

    static void setaPersonagem(String linha){
        String[] parte = linha.split(";");
        Personagem p = new Personagem();

        p.setId(parte[0]);
        p.setName(parte[1]);
        p.setAlternateNames(parte[2]);
        p.setHouse(parte[3]);
        p.setAncestry(parte[4]);
        p.setSpecies(parte[5]);
        p.setPatronus(parte[6]);
        p.setHogwartsStaff(parte[7]);
        p.setHogwartsStudent(parte[8]);
        p.setActorName(parte[9]);
        p.setAlive(parte[10]);
        p.setAlternateActors(parte[11]);
        p.setDateOfBirth(parte[12]);
        p.setYearOfBirth(parte[13]);
        p.setEyeColour(parte[14]);
        p.setGender(parte[15]);
        p.setHairColour(parte[16]);
        p.setWizard(parte[17]);

        Base[tamBase++]=p;
    }

    static void mostrarBase(){
        for (int i=0; i<tamBase; i++){
            Base[i].mostrarPersonagem();
        }
    }

    static void setaLista(){
        Scanner sc = new Scanner(System.in);
        String input = sc.nextLine();
        String chave = "";

        while(!input.equals("FIM")){
            setaInput(input);
            input = sc.nextLine();
        }

        chave = sc.nextLine();

        while(!chave.equals("FIM")){
            pesquisaSequencial(chave);
            chave = sc.nextLine();
        }

        sc.close();
    }

    static void setaInput(String id){
        for(int i=0; i<tamBase; i++){
            if(Base[i].getId().equals(id)){
                Lista[entradas++] = Base[i];
                i=tamBase;
            }
        }
    }

    static void mostrarLista(){
        for (int i=0; i<entradas; i++){
            Lista[i].mostrarPersonagem();
        }
    }

//////////////////////////////////////////////////////////////////////////////////////

    static void pesquisaSequencial(String x){
        boolean achou = false;

        for(int i=0; i<entradas && !achou; i++){
            if( (Lista[i].getName()).equals(x) ){
                System.out.println("SIM");
                achou=true;
            }
        }
        if(!achou) System.out.println("NAO");
    }

//////////////////////////////////////////////////////////////////////////////////////

    public static void main(String[] args){ 
        try {
            setaBase();
            setaLista();

        } catch(Exception e){ System.out.println(e);}
    }
}

File: Aeds/AedsII2/Trabalhos/Tp4/Q03.c
#include <time.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAXTAM 405

/**
 * Questao 3 - Trabalho Pratico 4 - Aeds II - CC_PUCMINAS
 * @author Tulio Gomes Braga
 * @version 1 06/2024
 * 
 * Explicações dos Métodos:
 *   Personagem: Responsavel pela cricao dos atributos dos personagens.
 *   mostraPersonagem: Apresenta na saida padrao os atributos formatados.
 *   mostraLista: Percorre toda a lista fazendo struct.mostrarPersonagem().
 *   mostrarBase: Percorre toda a base fazendo struct.mostrarPersonagem().
 *   setLocalDate: Recebe o parametro, trata, e o seta corretamente como Local Date.
 *   setaPersonagem: Particiona a linha e seta os atributos.
 *   setaLista: Percorre a base procurando o elemento compativel com chave de entrada, quando achar adiciona na lista.
 *   setaBase: Abre o arquivo uma vez e passa linha por linha para setar todos os personagens.
 *   lerEntrada: Enquanto a entrada for diferente de fim passa a chave pra ser procurada.
*/

typedef struct{
    int dia, mes, ano;
} LocalDate;

typedef struct{
    char id[50], name[50], alternate_names[200], house[50], ancestry[50], species[50], patronus[50], hogwartsStaff[50], hogwartsStudent[50], actorName[50], alternate_actors[100], eyeColour[50], gender[50], hairColour[50];
    LocalDate dateOfBirth;
    int yearOfBirth;
    bool alive, wizard; 
} Personagem;

typedef struct No{
    int nivel;
    struct No *esq, *dir;
    Personagem elemento;
}No;

No* novoNo(Personagem elemento){
    No *novo =  (No*)malloc(sizeof(No));
    novo->elemento = elemento;
    novo->esq = NULL;
    novo->dir = NULL;
    novo->nivel = 1;

    return novo;
}

Personagem Base[MAXTAM];
clock_t startTime, endTime;
int baseTam=0, entradas=0, comparacoes=0;
No *raiz = NULL;


//////////////////////////////////////////////////////// EXERCICIO 3

bool pesquisarNo(No *i, char *input){
    bool resp = false;

    if(i==NULL){ resp = false; comparacoes++;}
    else if(strcmp(input, i->elemento.name)<0){ printf(" esq"); resp = pesquisarNo(i->esq, input); comparacoes++;}
    else if(strcmp(input, i->elemento.name)>0){ printf(" dir"); resp = pesquisarNo(i->dir, input); comparacoes++;}
    else resp = true;
    
    return resp;
}

void pesquisar(){
    char input[100];
    scanf(" %[^\n\r]", input);

    while(strcmp(input, "FIM")!=0){
        printf("%s => raiz", input);
        printf("%s", pesquisarNo(raiz, input) ? " SIM\n" : " NAO\n");
        scanf(" %[^\n\r]", input);
    }  
}

int maxAbs(int x, int y){
    return x>y ? x : y;
}

int getNivel(No *i){
    return i==NULL ? 0 : i->nivel;
}

int setNivel(No* i){
    return 1 + maxAbs(getNivel(i->dir), getNivel(i->esq));
}

No* rotacionarEsq(No* i){
    No* noDir = i->dir;
    No* noDirEsq = noDir->esq;

    noDir->esq = i;
    i->dir = noDirEsq;

    i->nivel = setNivel(i);
    noDir->nivel = setNivel(noDir);
    return noDir;
}

No* rotacionarDir(No* i){
    No* noEsq = i->esq;
    No* noEsqDir = noEsq->dir;

    noEsq->dir = i;
    i->esq = noEsqDir;

    i->nivel = setNivel(i);
    noEsq->nivel = setNivel(noEsq);
    return noEsq;
}

No* balancear(No* i){
    if(i!=NULL){
        int fator = getNivel(i->dir) - getNivel(i->esq);

        if(fator>=-1 && fator<=1)  i->nivel = setNivel(i);
        else if(fator==2){
            int fatorFilhoDir = getNivel(i->dir->dir) - getNivel(i->dir->esq);
            if(fatorFilhoDir==-1) i->dir = rotacionarDir(i->dir);
            i = rotacionarEsq(i);
            comparacoes+=2;
        }else if(fator==-2){
            int fatorFilhoEsq = getNivel(i->esq->dir) - getNivel(i->esq->esq);
            if(fatorFilhoEsq==1) i->esq = rotacionarEsq(i->esq);
            i = rotacionarDir(i);
            comparacoes+=2;
        }
    }
    return i;
}

No* inserirNo(No* i, Personagem elemento){
    if(i==NULL){ i = novoNo(elemento); }
    else if(strcmp(elemento.name, i->elemento.name)<0){ i->esq = inserirNo(i->esq, elemento); comparacoes++; }
    else if(strcmp(elemento.name, i->elemento.name)>0){ i->dir = inserirNo(i->dir, elemento); comparacoes++; }
    return balancear(i);
}

void inserir(Personagem elemento){
    raiz = inserirNo(raiz, elemento);
}

void Arqlog(){
    double ticks = (double)(endTime - startTime);    // Ticks de clock da cpu (unidades clock, "pulsos")
    double duracaoDeUmTick = 1.0 / CLOCKS_PER_SEC;  // Duração de um tick em segundos
    double tempo = ticks * duracaoDeUmTick;

    FILE *file = fopen("802512_avl.txt", "w");
    fprintf(file, "802512\t%i\t%.6fs", comparacoes, tempo);
}

////////////////////////////////////////////////////////////////////


void mostraPersonagem(Personagem p, int i){
    printf("[%i ## %s ## %s ## {%s} ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %02i-%02i-%i ## %i ## %s ## %s ## %s ## %s]\n" , 
        i,
        p.id, 
        p.name, 
        p.alternate_names,
        p.house,
        p.ancestry,
        p.species,
        p.patronus,
        p.hogwartsStaff,
        p.hogwartsStudent,
        p.actorName,
        p.alive ? "true" : "false",
        p.dateOfBirth.dia, p.dateOfBirth.mes, p.dateOfBirth.ano,
        p.yearOfBirth,
        p.eyeColour,
        p.gender,
        p.hairColour,
        p.wizard ? "true" : "false"
    );
}

void mostrarBase(){
    for(int i=0; i<baseTam; i++){
        mostraPersonagem(Base[i], i);
    }
}

void setaLocalDate(Personagem *p, char data[]){ 
    char dia[3], mes[3], ano[5];

    strncpy(dia, data, 2);
    dia[2] = '\0'; 
    p->dateOfBirth.dia = atoi(dia);

    if(data[4]!='-'){   strncpy(mes, data+3, 2);    mes[2]='\0';    p->dateOfBirth.mes = atoi(mes);}
    else{ strncpy(mes, data+3, 1);    mes[1]='\0';    p->dateOfBirth.mes = atoi(mes);}

    if(data[4]!='-'){   strncpy(ano, data+6, 4);    ano[4]='\0';    p->dateOfBirth.ano = atoi(ano);}
    else{ strncpy(ano, data+5, 4);  ano[4]='\0';    p->dateOfBirth.ano = atoi(ano);}
}

void setaPersonagem(char* linha){
    char input[300], data[11]; strcpy(input,linha);
    int x=0, y=0, l=0;

    while(input[x]!=';'){ Base[baseTam].id[y] = input[x]; x++; y++;}    Base[baseTam].id[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].name[y] = input[x]; x++; y++;}  Base[baseTam].name[y]='\0'; x++; y=0;
    while(input[x]!=';'){ if(input[x]!='[' && input[x]!=']' && input[x]!=39){ Base[baseTam].alternate_names[y] = input[x]; y++;} x++;}  Base[baseTam].alternate_names[y]='\0'; x++; y=0;   
    while(input[x]!=';'){ Base[baseTam].house[y] = input[x]; x++; y++;} Base[baseTam].house[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].ancestry[y] = input[x]; x++; y++;}  Base[baseTam].ancestry[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].species[y] = input[x]; x++; y++;}   Base[baseTam].species[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].patronus[y] = input[x]; x++; y++;}  Base[baseTam].patronus[y]='\0'; x++; y=0;
    while(input[x]!=';'){ if(input[x]=='F'){ strcpy(Base[baseTam].hogwartsStaff,"false"); x+=5;} else{strcpy(Base[baseTam].hogwartsStaff,"true"); x+=10;}} x++; y=0;
    while(input[x]!=';'){ if(input[x]=='F'){ strcpy(Base[baseTam].hogwartsStudent,"false"); x+=5;} else{strcpy(Base[baseTam].hogwartsStudent,"true"); x+=10;}} x++; y=0;
    while(input[x]!=';'){ Base[baseTam].actorName[y] = input[x]; x++; y++;}  Base[baseTam].actorName[y]='\0'; x++; y=0;
    while(input[x]!=';'){ if(input[x]=='V'){ Base[baseTam].alive=true; x+=10;} else{ Base[baseTam].alive=false; x+=5;}} x++; y=0;
    while(input[x]!=';'){ Base[baseTam].alternate_actors[y] = input[x]; x++; y++;}  Base[baseTam].alternate_actors[y]='\0'; x++; y=0;
    while(input[x]!=';'){ data[y] = input[x]; x++; y++;} setaLocalDate(&Base[baseTam],data); x++; y=0;
    while(input[x]!=';'){ data[y] = input[x]; x++; y++;} data[y]='\0'; Base[baseTam].yearOfBirth = atoi(data); x++; y=0;
    while(input[x]!=';'){ Base[baseTam].eyeColour[y] = input[x]; x++; y++;}  Base[baseTam].eyeColour[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].gender[y] = input[x]; x++; y++;}  Base[baseTam].gender[y]='\0'; x++; y=0;
    while(input[x]!=';'){ Base[baseTam].hairColour[y] = input[x]; x++; y++;}  Base[baseTam].hairColour[y]='\0'; x++; y=0;
    if(input[x]=='V'){ Base[baseTam].wizard=true; x+=9;} else{ Base[baseTam].wizard=false; x+=4;}

    baseTam++;
}

void setaArvore(char input[]){
    bool achou = false;

    for(int i=0; i<MAXTAM && !achou; i++){
        if(strcmp(Base[i].id, input)==0){
            inserir(Base[i]);
            achou = true;
        }
    }
}

void setaEntrada(){
    char input[100];
    scanf("%s", input);

    while(strcmp(input, "FIM")!=0){
        setaArvore(input);
        scanf("%s", input);
    }
}

void setaBase(){
    FILE *file =  fopen("/tmp/characters.csv","r");
    char input[300];
    fgets(input, sizeof(input), file);
    
    while(fgets(input, sizeof(input), file) != NULL){
        setaPersonagem(input);
    }
}

void mostrarNo(No *i) {
    if (i != NULL) {
        mostrarNo(i->esq);
        mostraPersonagem(i->elemento, 0);
        mostrarNo(i->dir);
    }
}


int main(){
    startTime = clock();
    setaBase();
    setaEntrada();
    //mostrarNo(raiz);
    pesquisar();
    endTime = clock();

    Arqlog();
}